geoSegment.js

```diff
@@ -0,0 +1,353 @@
+  // generation algorithm below.
+  byDay: Joi.array().items(Joi.string()).min(1).required(),
+  startTime: Joi.string().required(),
+  endTime: Joi.string().required(),
+  idTemplate: Joi.string().required(),
+});
+
+/**
+ * @typedef {import('axios').AxiosError} AxiosError
+ * @typedef {import('axios').AxiosRequestConfig} AxiosRequestConfig
+ * @typedef {import('axios').AxiosResponse} AxiosResponse
+ * @typedef {{
+ * location?: {
+ *  geo?: {
+ *   latitude: number,
+ *   longitude: number,
+ *   }
+ *  },
+ * [k:string]: unknown
+ * }} SessionSeriesData
+ * @typedef {{
+ *  id: string,
+ *  state: 'updated' | 'deleted',
+ *  kind: 'SessionSeries',
+ *  data: SessionSeriesData
+ * }} SessionSeriesItem
+ * @typedef {{
+ *  startDate?: string,
+ *  superEvent?: string,
+ *  [k:string]: unknown
+ * }} ScheduledSessionData
+ * @typedef {{
+ *  id: string,
+ *  state: 'updated' | 'deleted',
+ *  kind: 'ScheduledSession',
+ *  data: ScheduledSessionData
+ * }} ScheduledSessionItem
+ * @typedef {{
+ *  latitude: number,
+ *  longitude: number,
+ *  radius: number,
+ *  identifier: string
+ * }} Segment
+ */
+
+/**
+ * @param {Segment[]} segments
+ */
+async function createSegmentDirectories(segments) {
+  // Create segment directories;
+  segments.map((segment) => segment.identifier)
+    .map(async (segmentIdentifier) => {
+      const segmentDirectoryPath = `${OUTPUT_DIRECTORY_PATH}/${segmentIdentifier}`;
+      await emptyOrMakeDirectory(segmentDirectoryPath);
+
+      // Create index file
+      await fsPromises.writeFile(`${OUTPUT_DIRECTORY_PATH}/${segmentIdentifier}/index.txt`, '');
+    });
+}
+
+/**
+ * @param {ScheduledSessionData} scheduledSessionData
+ * @param {SessionSeriesData} sessionSeries
+ */
+async function linkScheduledSessionAndSessionSeriesAndWrite(scheduledSessionData, sessionSeries) {
+  // Link ScheduledSession with SessionSeries
+  const linkedScheduledSessionData = {
+    ...scheduledSessionData,
+    superEvent: sessionSeries,
+  };
+
+  // Write ScheduledSession into each output segment directory
+  for (const segmentIdentifier of sessionSeries['imin:segment']) {
+    const filename = `${encodeURIComponent(linkedScheduledSessionData.id)}.json`;
+    await fs.writeJson(`${OUTPUT_DIRECTORY_PATH}/${segmentIdentifier}/${filename}`, linkedScheduledSessionData);
+    await fs.appendFile(`${OUTPUT_DIRECTORY_PATH}/${segmentIdentifier}/index.txt`, `${filename}\r\n`);
+  }
+}
+
+/**
+ * There is a system filepath limit of 255 characters in most file systems.
+ * All sessionseries filepaths with be of the form `./sessionseries/${encodedFilename}.json`
+ * Therefore the encodedFilename has to be less than 234 characters in length (./sessionseries/.json is 21 characters)
+ *
+ * @param {string} sessionSeriesId
+ */
+function getSessionSeriesFilePath(sessionSeriesId) {
+  let encodedFilename = encodeURIComponent(sessionSeriesId);
+  if (encodedFilename.length > 234) {
+    encodedFilename = encodedFilename.slice(0, 234);
```

Comment:
This a great shout! Should we slice the last characters rather than the first (as the specificity of the path is likely at the end of the string?)

Reply:
Though actually hashing is probably the way to go - likely more robust and repeatable across environments (the risk is then collisions, which we can test and control for in our own environment, rather than the risk of filesystem issues which is much harder to control for and test properly)